lib/src/client/dio_interceptor.dart

```diff
@@ -0,0 +1,198 @@
+        print('[✓] onResponse triggered');
+
+        _clockSkew = serverDate - now;
+        // Re-sign the request with the updated clockskew
+        _signRequest(response.requestOptions);
+        // Resend the request
+        _resendRequest(response.requestOptions, handler);
+      }
+    }
+    return handler.next(response);
+  }
+
+  Future<void> _resendRequest(
+      RequestOptions options, ResponseInterceptorHandler handler) async {
+    print('[✓] resendRequest triggered');
+    final response = await _dioInstance.fetch(options);
+    return handler.next(response);
+  }
+
+  /// Signs the outgoing request with a request signature.
+  Future<void> _signRequest(RequestOptions options) async {
+    print('[✓] _signRequest triggered');
+    options.headers.remove(_options.headerName);
+
+    final signatureBodySourceComponents =
+        _options.signatureBodySourceComponents.isNotEmpty
+            ? _options.signatureBodySourceComponents
+            : DefaultConstants.signatureBodySourceComponents;
+
+    Uint8List? body;
+    // Encode request body if required by signature components
+    if (options.data != null &&
+        signatureBodySourceComponents
+            .contains(SignatureBodySourceComponents.body)) {
+      body = utf8.encode(options.data.toString());
+    }
+
+    // Build parameters for constructing the signature body
+    final signatureBodySourceParameters = SignatureBodySourceParameters(
+        options.method,
+        options.uri,
+        options.headers
+            .map((key, value) => MapEntry(key.toString(), value.toString())),
+        _uuid.v4(), // Generate a nonce
+        _getTimestamp(options),
+        _options.clientId!,
+        signatureBodySourceComponents,
+        body: body);
+
+    // Build the signature body source based on parameters
+    final signatureBodySource =
+        await _signatureBodySourceBuilder.build(signatureBodySourceParameters);
+
+    // Construct parameters for creating the signature
+    final signatureBodyParameters =
+        SignatureBodyParameters(signatureBodySource, _options.clientSecret!);
+
+    // Generate the signature using the signer
+    final signatureBody =
+        await _signatureBodySigner.sign(signatureBodyParameters);
+
+    // Format the signature header based on the specified pattern
+    final signatureHeader = _options.signaturePattern
+        .replaceAll("{ClientId}", _options.clientId!)
+        .replaceAll("{Nonce}", signatureBodySourceParameters.nonce)
+        .replaceAll(
+            "{Timestamp}", signatureBodySourceParameters.timestamp.toString())
+        .replaceAll("{SignatureBody}", signatureBody);
+
+    print('[DEBUG] Generated Signature Header: $signatureHeader');
+
+    // Add the signature header to the request headers
+    options.headers[_options.headerName] = signatureHeader;
+  }
+
+  // Returns what the client thinks is the current time.
+  int getTime(RequestOptions request) {
```

Comment:
This method seems to be the homologue of the `GetTime` method of the .NET `RequestsSignatureDelegatingHandler` (because it has the same name and the same comment). However, the behavior is completely different.
- The .NET `GetTime` gets the time based on either `DateTimeOffset.UtcNow` or an injected function.
- This Dart `getTime` gets the time by parsing the value of the `X-Signature` header, which I don't really understand how it would work because we're in the process of writing that header. Moreover, you assume the timestamp is at index 2 of the signature, but the signature pattern is actually customizable so you don't have that guarantee.

Reply:
the last commit should patch this one up